doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,100 +820,189 @@ tactic
+
+      :tacn:`match` :n:`…` is, in fact, shorthand for :tacn:`once` :tacn:`multimatch` `…`.
+
+   `lazymatch`
+      Causes the match to commit to the first matching branch
+      rather than trying a new match if :token:`ltac_expr` fails. If
+      the :token:`ltac_expr` of the selected branch is a tactic with
+      backtracking points, then subsequent failures cause this tactic to
+      backtrack.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of terms, but it is extended with pattern matching
+      metavariables. In :token:`cpattern`, a pattern matching metavariable is
+      represented with the syntax :n:`?@ident`. The
+      notation :g:`_` can also be used to denote metavariable whose instance is
+      irrelevant. In the notation :n:`?@ident`, the identifier allows us to keep
+      instantiations and to make constraints whereas :g:`_` shows that we are not
+      interested in what will be matched. On the right hand side of pattern matching
+      clauses, the named metavariables are used without the question mark prefix. There
+      is also a special notation for second-order pattern matching: in an
+      applicative pattern of the form :n:`@`\ :n:`?@ident @ident__1 … @ident__n`,
+      the variable :token:`ident` matches any complex expression with (possible)
+      dependencies in the variables :n:`@ident__i` and returns a functional term
+      of the form :n:`fun @ident__1 … @ident__n => @term`.
+
+   :n:`context {? @ident } [ @cpattern ]`
+      This special form of pattern matches any term with a subterm matching
+      cpattern. If there is a match, the optional :n:`@ident` is assigned the "matched
+      context", i.e. the initial term where the matched subterm is replaced by a
+      hole. The example below will show how to use such term contexts.
+
+      If the evaluation of the :token:`ltac_expr` of a valid match fails, the next
+      matching subterm is tried. If no further subterm matches, the next clause
+      is tried. Matching subterms are considered from top to bottom and from left to
+      right (with respect to the raw printing obtained by setting the
+      :flag:`Printing All` flag).
+
+     .. example:: Use of `context` in match on terms
+
+        .. coqtop:: all abort
 
-We can carry out pattern matching on terms with:
+           Ltac f x :=
+             match x with
+               context f [S ?X] =>
+               idtac X;                    (* To display the evaluation order *)
+               assert (p := eq_refl 1 : X=1);    (* To filter the case X=1 *)
+               let x:= context f[O] in assert (x=O) (* To observe the context *)
+             end.
+           Goal True.
+           f (3+4).
 
-.. tacn:: match @ltac_expr with {+| @cpattern__i => @ltac_expr__i} end
+   .. todo @term after context is actually @lconstr_pattern (term with holes), which is spliced away
+      Maybe worth restoring that distinction?
```

Comment:
I don't understand this, what is `@term`.  We seem to be using `@cpattern`?  Also, where is the syntax for `let @ident := context @ident [ @term ] in ...` explained / defined?

Reply:
`cpattern` is a synonym for `term` to help with the presentation.  We use `term` in the doc instead of `lconstr`.  It's equivalent to `operconstr` level 200 in the mlgs.

You should look at the HTML version of the doc.  Everything in the grammar is hyperlinked.